### PR TITLE
Expose and test new String(ReadOnlySpan<char>) ctor

### PR DESCRIPTION
**Pull request by stephentoub**
cc: @jkotas, @KrzysztofCwalina 
Contributes to https://github.com/dotnet/corefx/issues/22378
Depends on https://github.com/dotnet/coreclr/pull/13075 being merged and consumed into corefx before this PR will pass CI.

### CODE REVIEW

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
@dotnet-bot test this please

**Comment by jkotas** [comment]
@MichalStrehovsky Could you please push CoreRT master to nmirror when you get a chance to get this unblocked?

**Comment by MichalStrehovsky** [comment]
> @MichalStrehovsky Could you please push CoreRT master to nmirror when you get a chance to get this unblocked?

Don't hold your breath:

```
NUTC3039:Internal Compiler Error: Field 'BoxedValue' on type 'System.__Boxed`1<System.ReadOnlySpan`1<T_System.__UniversalCanon>>' from assembly 'System.Private.CoreLib' is a byref-like field that is not defined on a byref-like type while loading type 'System.__Boxed`1<System.ReadOnlySpan`1<T_System.__UniversalCanon>>'. while compiling method 'instance System.Void System.ReadOnlySpan`1<System.__UniversalCanon>..ctor_UnboxingStub(System.__UniversalCanon[], System.Int32)'
```

**Comment by MichalStrehovsky** [comment]
> Don't hold your breath

Turns out our mainline branch on the TFS side was broken, and the CoreCLR changes turned out to be problem-free. I'm kicking off a master -> nmirror integration now.

(But we did stop CoreFX -> ProjectN integrations because that's where the problem breaking TFS mainline came from over the weekend).

**Comment by stephentoub** [comment]
> But we did stop CoreFX -> ProjectN integrations because that's where the problem breaking TFS mainline came from over the weekend

Just for reference, which corefx change broke it?

**Comment by MichalStrehovsky** [comment]
`Microsoft.NETCore.UniversalWindowsPlatform 6.0.0-preview2-25527-02` is known to work, `6.0.0-preview2-25530-02` is known to be broken for the Project N compiler. I didn't try to chase it down to a pull request.

**Comment by jkotas** [comment]
@dotnet-bot test this please

**Comment by stephentoub** [comment]
@dotnet-bot test Linux arm Release Build please ("Unable to find image", fyi @dotnet/dnceng)

File: src/System.Runtime/tests/System/StringTests.netcoreapp.cs [hunk 1]
```diff
@@ -12,6 +12,34 @@ namespace System.Tests
...
+        [Theory]
+        [InlineData(new char[] { 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', '\0' }, 0, 8, "abcdefgh")]
+        [InlineData(new char[] { 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', '\0' }, 0, 9, "abcdefgh\0")]
+        [InlineData(new char[] { 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', '\0', 'i', 'j', 'k' }, 0, 12, "abcdefgh\0ijk")]
+        [InlineData(new char[] { 'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', '\0' }, 2, 3, "cde")]
+        [InlineData(new char[] { '\0' }, 0, 1, "\0")]
+        [InlineData(new char[] { 'a', 'b', 'c' }, 0, 0, "")]
+        [InlineData(new char[] { 'a', 'b', 'c' }, 1, 0, "")]
+        [InlineData(new char[] { 'a', 'b', 'c' }, 0, 1, "a")]
+        [InlineData(new char[] { 'a', 'b', 'c' }, 1, 1, "b")]
```
Comment (jkotas) [inline src/System.Runtime/tests/System/StringTests.netcoreapp.cs]: Nit: This feels like way too many redundant test cases, not exercising any unique paths. I think this list can be shorter.

Comment (stephentoub) [inline src/System.Runtime/tests/System/StringTests.netcoreapp.cs]: I've removed a bunch of the cases.

**Comment by jkotas** [comment]
👍

**Pull request merged**